### ISSUE TITLE
Error with unicorn/rainbow when adding dictionary items

### ISSUE THREAD

**Issue by devstack** [kamsar/Unicorn#98]
Using the following versions:

Rainbow version 1.1.0-beta03
Rainbow.Core version 1.1.0-beta03
Rainbow.Storage.Sc version 1.1.0-beta03
Rainbow.Storage.Yaml version 1.1.0-beta03
Unicorn version 3.1.0-beta04
Unicorn.Core version 3.1.0-beta04

When selecting Add New Dictionary Entry in sitecore content tree receive the following error:

[ArgumentNullException: Value cannot be null.
Parameter name: source]
   System.Linq.Enumerable.Select(IEnumerable`1 source, Func`2 selector) +5974671
   Rainbow.Model.ProxyItem..ctor(IItemData itemToProxy) +605
   Rainbow.Storage.SerializationFileSystemTree.WriteItem(IItemData item, String path) +122
   Rainbow.Storage.SerializationFileSystemTree.Save(IItemData item) +72
   Unicorn.Data.DataProvider.UnicornDataProvider.SerializeItemIfIncluded(IItemData item, String triggerReason) +101
   Unicorn.Data.DataProvider.UnicornSqlServerDataProvider.CreateItem(ID itemId, String itemName, ID templateId, ItemDefinition parent, CallContext context) +212
   Sitecore.Data.DataProviders.DataProvider.CreateItem(ID itemID, String itemName, ID templateID, ItemDefinition parent, CallContext context, DataProviderCollection providers) +159
   Sitecore.Data.DataSource.CreateItem(ID itemID, String itemName, ID templateID, ID parentID) +104
   Sitecore.Nexus.Data.DataCommands.CreateItemCommand.Execute(ID itemId, String itemName, ID templateId, Item destination) +500
   Sitecore.Data.Engines.DataCommands.CreateItemCommand.DoExecute() +109
   Sitecore.Data.Engines.EngineCommand`2.Execute() +96
   Sitecore.Data.Managers.ItemProvider.DoCreateAndReturnItem(String itemName, Item destination, ID templateId, ID newId) +69
   Sitecore.Data.Managers.ItemProvider.CreateItem(String itemName, Item destination, ID templateId, ID newId, SecurityCheck securityCheck) +362
   Sitecore.Data.Managers.<>c__DisplayClass13.<CreateItem>b__12() +58
   Sitecore.Data.Managers.PipelineBasedItemProvider.ExecuteAndReturnResult(String pipelineName, String pipelineDomain, Func`1 pipelineArgsCreator, Func`1 fallbackResult) +96
   Sitecore.Data.Managers.ItemManager.CreateItem(String itemName, Item destination, ID templateId, ID newId, SecurityCheck securityCheck) +147
   Sitecore.Data.Managers.ItemManager.CreateItem(String itemName, Item destination, ID templateId, ID newId) +127
   Sitecore.Nexus.Data.DataCommands.AddFromTemplateCommand.(String , Item , ID , ID ) +11
   Sitecore.Data.Engines.DataCommands.AddFromTemplateCommand.DoExecute() +116
   Sitecore.Data.Engines.EngineCommand`2.Execute() +96
   Sitecore.Data.Managers.ItemProvider.AddFromTemplate(String itemName, ID templateId, Item destination, ID newId) +375
   Sitecore.Data.Managers.<>c__DisplayClass3.<AddFromTemplate>b__2() +51
   Sitecore.Data.Managers.PipelineBasedItemProvider.ExecuteAndReturnResult(String pipelineName, String pipelineDomain, Func`1 pipelineArgsCreator, Func`1 fallbackResult) +96
   Sitecore.Data.Managers.ItemManager.AddFromTemplate(String itemName, ID templateId, Item destination, ID newItemId) +136
   Sitecore.Data.Managers.ItemManager.AddFromTemplate(String itemName, ID templateId, Item destination) +155
   Sitecore.Data.Items.Item.Add(String name, TemplateID templateID) +83
   Sitecore.Workflows.WorkflowContext.AddItem(String name, TemplateItem template, Item parent) +79
   Sitecore.Shell.Framework.Commands.AddMaster.Add(ClientPipelineArgs args) +906

[TargetInvocationException: Exception has been thrown by the target of an invocation.]
   System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor) +0
   System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments) +128
   System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) +146
   Sitecore.Reflection.ReflectionUtil.InvokeMethod(MethodInfo method, Object[] parameters, Object obj) +89
   Sitecore.Nexus.Pipelines.NexusPipelineApi.Resume(PipelineArgs args, Pipeline pipeline) +301
   Sitecore.Web.UI.Sheer.ClientPage.ResumePipeline() +224
   Sitecore.Web.UI.Sheer.ClientPage.OnPreRender(EventArgs e) +779
   Sitecore.Shell.Applications.ContentManager.ContentEditorPage.OnPreRender(EventArgs e) +24
   System.Web.UI.Control.PreRenderRecursiveInternal() +107
   System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint) +4171

**Comment by devstack** [comment]
This appears to be a bigger issue - having problems creating items in general in the sitecore tree.

**Comment by kamsar** [comment]
Hi @devstack have you tried upgrading to the latest 3.1 beta? I've made a number of fixes since 3.1b4, the latest is now 3.1b7

**Comment by kamsar** [comment]
I was not able to reproduce on 3.1b7 - I'm pretty sure that this was an issue that I found back around 3.1b5 and fixed.

**Comment by devstack** [comment]
I'm using
C:\src\VCWebsite\src\packages\Unicorn.Core.3.1.0-beta07\lib\net45\Unicorn.dll
and do get the issue.

[image: Inline image 1]

On Mon, Dec 21, 2015 at 5:02 PM, Kam Figy <notifications@github.com> wrote:

> I was not able to reproduce on 3.1b7 - I'm pretty sure that this was an
> issue that I found back around 3.1b5 and fixed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kamsar/Unicorn/issues/98#issuecomment-166359999>.
>



-- 

*Tom Miller*
Director
Devstack Consultancy Ltd
tom@devstack.co.uk
+44 7810 387300

**Comment by GuitarRich** [comment]
I have the same issue on 3.1b7 - I'll see if I can track it down. @kamsar what was the issue in 3.1b5 that was fixed? Maybe its related?

**Comment by kamsar** [comment]
It is likely one of the collections on the item being proxied, unversioned
fields maybe, being null. Odd that I can't reproduce. Will look later. Am
on vacation at the moment :)

**Comment by GuitarRich** [comment]
The error is occurring on this line in `ProxyItem.cs`
```
UnversionedFields = itemToProxy.UnversionedFields.Select(language => new ProxyItemLanguage(language)).ToArray();
```

For some reason the `itemToProxy.UnversionedFields` is null. Trying to work out why, if its a bad thing and how to cope with it.

**Comment by kamsar** [comment]
It's probably a bug on the default value. I noticed that it came from
CreateItem() in the stack trace, may well be that the constructor for the
proxy cooked up in there is leaving the collection null when it should be
empty.

Thanks for helping debug!

**Comment by GuitarRich** [comment]
Thats the bad boy!  The constructor in `ProxyItem.cs` that sets the Field enumerables to empty lists is not setting the `UnversionedFields` property

```
		public ProxyItem(string name, Guid id, Guid parentId, Guid templateId, string path, string databaseName)
		{
			Assert.ArgumentNotNullOrEmpty(name, "name");
			Assert.ArgumentNotNullOrEmpty(path, "path");
			Assert.ArgumentNotNullOrEmpty(databaseName, "databaseName");

			Name = name;
			Id = id;
			ParentId = parentId;
			TemplateId = templateId;
			Path = path;
			DatabaseName = databaseName;
			SharedFields = Enumerable.Empty<IItemFieldValue>();
			Versions = Enumerable.Empty<IItemVersion>();
		}
``` 

Looks like this should be added to that:

```

**Comment by GuitarRich** [comment]
I'll submit a pull request for it :)

**Comment by GuitarRich** [comment]
Technically this is an issue with Rainbow, not Unicorn. Also looking at the tests on Rainbow, I think the constructor in `FakeItem` is masking the issue by setting the property to an empty list if it is null

```csharp
public FakeItem(
	Guid id = default(Guid), 
	string databaseName = "master", 
	Guid parentId = default(Guid), 
	string path = "/sitecore/content/test item", 
	string name = "test item", 
	Guid branchId = default(Guid), 
	Guid templateId = default(Guid), 
	IEnumerable<IItemFieldValue> sharedFields = null, 
	IEnumerable<IItemVersion> versions = null, 
	string serializedItemId = "0xDEADBEEF", 
	IEnumerable<IItemData> children = null,
	IEnumerable<IItemLanguage> unversionedFields = null) : base(name, id, parentId, templateId, path, databaseName)
{
	BranchId = branchId;
	SharedFields = sharedFields ?? new List<IItemFieldValue>();
	Versions = versions ?? new List<IItemVersion>();
	SerializedItemId = serializedItemId;
	UnversionedFields = unversionedFields ?? new List<IItemLanguage>();

	SetProxyChildren(children?.ToArray() ?? new IItemData[0]);
}
```

**Comment by GuitarRich** [comment]
Pull request created, https://github.com/kamsar/Rainbow/pull/9 - not sure what to do about the unit tests because of the above comment. @kamsar when you get back, let me know what you think. Enjoy your vacation

**Comment by kamsar** [comment]
PR accepted, thank you. Rainbow 1.1b5 has been pushed to NuGet with the fix.

**Comment by GuitarRich** [comment]
Wow - that was fast :+1: Awesome, thanks!

**Comment by devstack** [comment]
Thanks both for helping out with this!

On Wed, Dec 23, 2015 at 7:59 PM, Richard Seal <notifications@github.com>
wrote:

> Wow - that was fast [image: :+1:] Awesome, thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kamsar/Unicorn/issues/98#issuecomment-166980428>.
>



-- 

*Tom Miller*
Director
Devstack Consultancy Ltd
tom@devstack.co.uk
+44 7810 387300

**Comment by kamsar** [comment]
This is fixed in 3.1 final.
